### ISSUE TITLE
Verify of certificate chain using RSA-PSS keys fails starting with 1.1.1h

### ISSUE THREAD

**Issue by gingerbeard-man** [openssl/openssl#13931]
We have a two-level CA using RSA-PSS keys throughout.  On a system with OpenSSL 1.1.1h verification of this CA's end-entity certificates fails with "lookup: unable to get local issuer certificate".  The same certificate chain verifies OK with older releases up to and including 1.1.1g.  OpenSSL 1.1.1i and 3.0.0-alpha1 also report the same error.

I can reproduce this issue with fresh builds of 1.1.1g and 1.1.1h on Linux using the following CONF files and commands:

```
# rootCA.cnf

[ req ]
prompt              = no
distinguished_name  = req_distinguished_name
x509_extensions     = usr_cert

[ req_distinguished_name ]
C  = DE
O  = Test Org
CN = Test RSA PSS Root-CA

[ usr_cert ]
basicConstraints        = critical,CA:TRUE
keyUsage                = critical,keyCertSign,cRLSign
subjectKeyIdentifier    = hash
authorityKeyIdentifier  = keyid:always
```

```
# subCA.cnf

[ req ]
prompt              = no
distinguished_name  = req_distinguished_name
x509_extensions     = usr_cert

[ req_distinguished_name ]
C  = DE
O  = Test Org
CN = Test RSA PSS Sub-CA

[ usr_cert ]
basicConstraints        = critical,CA:TRUE,pathlen:0
keyUsage                = critical,keyCertSign,cRLSign
subjectKeyIdentifier    = hash
authorityKeyIdentifier  = keyid:always
```

```
# user.cnf

[ req ]
prompt              = no
distinguished_name  = req_distinguished_name
x509_extensions     = usr_cert

[ req_distinguished_name ]
C  = DE
O  = Test Org
CN = Test User

[ usr_cert ]
keyUsage                = critical,digitalSignature,keyAgreement
extendedKeyUsage        = clientAuth,serverAuth
subjectKeyIdentifier    = hash
authorityKeyIdentifier  = keyid:always
```

```
$ openssl genpkey -algorithm RSA-PSS -out rootCA_key.pem -pkeyopt rsa_keygen_bits:2048
$ openssl req -config rootCA.cnf -set_serial 01 -new -batch -sha256 -nodes -x509 -days 9125 -out CA/rootCA_cert.pem -key rootCA_key.pem -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1

$ openssl genpkey -algorithm RSA-PSS -out subCA_key.pem -pkeyopt rsa_keygen_bits:2048
$ openssl req -config subCA.cnf -new -out subCA_req.pem -key subCA_key.pem -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1
$ openssl x509 -req -sha256 -in subCA_req.pem -CA CA/rootCA_cert.pem -CAkey rootCA_key.pem -out CA/subCA_cert.pem -CAserial rootCA_serial.txt -CAcreateserial -extfile subCA.cnf -extensions usr_cert -days 4380 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1
$ c_rehash CA

$ openssl genpkey -algorithm RSA-PSS -out user1_key.pem -pkeyopt rsa_keygen_bits:2048
$ openssl req -config user.cnf -new -out user1_req.pem -key user1_key.pem -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1
$ openssl x509 -req -sha256 -in user1_req.pem -CA CA/subCA_cert.pem -CAkey subCA_key.pem -out user1_cert.pem -CAserial subCA_serial.txt -CAcreateserial -extfile user.cnf -extensions usr_cert -days 1825 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1
```

Verify with OpenSSL 1.1.1g:

```
$ openssl version
OpenSSL 1.1.1g  21 Apr 2020
$ openssl verify -show_chain -CApath CA user1_cert.pem
user1_cert.pem: OK
Chain:
depth=0: C = DE, O = Test Org, CN = Test User (untrusted)
depth=1: C = DE, O = Test Org, CN = Test RSA PSS Sub-CA
depth=2: C = DE, O = Test Org, CN = Test RSA PSS Root-CA
```

Verify with OpenSSL 1.1.1h:

```
$ openssl version
OpenSSL 1.1.1h  22 Sep 2020
$ openssl verify -show_chain -CApath CA user1_cert.pem
C = DE, O = Test Org, CN = Test User
error 20 at 0 depth lookup: unable to get local issuer certificate
error user1_cert.pem: verification failed
```

If I generate the intermediate certificate with a plain RSA key instead of RSA-PSS then the error occurs one level higher:

```
$ openssl verify -show_chain -CApath CA user1_cert.pem
C = DE, O = Test Org, CN = Test RSA PSS Sub-CA
error 2 at 1 depth lookup: unable to get issuer certificate
error user1_cert.pem: verification failed
```

Using plain RSA for both CA certificates the chain verifies OK with all OpenSSL versions.

**Comment by russhousley** [comment]
I see the same error on MacOS 10.14.6 with OpenSSL 1.1.1.h.

**Comment by mattcaswell** [comment]
This seems to have been caused by commit e2590c3a1 from PR #10587.

Perhaps @DDvO can comment.

**Comment by richsalz** [comment]
Will any fix also need to be made in master?

**Comment by mattcaswell** [comment]
I can recreate the same verify failure in master. So if this is confirmed as a bug (not yet clear that it is), then the fix will need to be made in master too.

Interestingly the steps for creating the chain given above fail on master at this stage:

````
$ openssl req -config subCA.cnf -new -out subCA_req.pem -key subCA_key.pem -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1
80E213F7BE7F0000:error:0300008B:digital envelope routines:EVP_DigestInit_ex:no digest set:crypto/evp/digest.c:173:
````

Which looks like a different, independent problem.

**Comment by mattcaswell** [comment]
> Interestingly the steps for creating the chain given above fail on master at this stage:

Raised as a separate issue: #13957

**Comment by DDvO** [comment]
Things go wrong here:
```
/* Check that issuer public key algorithm matches subject signature algorithm */
static int check_sig_alg_match(const EVP_PKEY *pkey, const X509 *subject)
{
    int pkey_nid;

    if (pkey == NULL)
        return X509_V_ERR_NO_ISSUER_PUBLIC_KEY;
    if (OBJ_find_sigid_algs(OBJ_obj2nid(subject->cert_info.signature.algorithm),
                            NULL, &pkey_nid) == 0)
        return X509_V_ERR_UNSUPPORTED_SIGNATURE_ALGORITHM;
    if (EVP_PKEY_type(pkey_nid) != EVP_PKEY_base_id(pkey))
        return X509_V_ERR_SIGNATURE_ALGORITHM_MISMATCH;
```
In the RSA-PSS case, 
`EVP_PKEY_type(pkey_nid)` is 6 (rsaEncryption) while 
`EVP_PKEY_base_id(pkey)` is 912 (rsassaPss).

**Comment by DDvO** [comment]
Is the mapping `rsassaPss		undef	rsaEncryption` in `obj_xref.txt` correct?

**Comment by DDvO** [comment]
A related issue is that in case chain building goes wrong the diagnostics is often too general (such as "unable to get local issuer certificate") and thus misleading.
In this case, showing "signature algorithm mismatch" would be much more to the point (though in this case due to some bug).

**Comment by DDvO** [comment]
The actual problem is that RSA public keys in CA certs can be marked simply as "RSA" or "RSA-PSS", and this may match or not with RSA-PSS signatures of certs they sign.

**Comment by DDvO** [comment]
So a solution is to weaken the comparison to use the base signature type of both the pkey and the subject cert, which is RSA here.

I've provided PR #13968 (for the master)
which also adds a test that is a little more challenging than the example given above,
using a plain RSA root CA cert and an RSA-PSS issuing CA cert.

**Comment by DDvO** [comment]
BTW, the so far not general enough sig ald matching goes back to commit ffd2df135a5d9f6d2627bd125f362298430fdc06
but its way of affecting chain building changed later, maybe due my improvements done last year regarding the fix of #1418.

**Comment by t8m** [comment]
> BTW, the so far not general enough sig ald matching goes back to commit [ffd2df1](https://github.com/openssl/openssl/commit/ffd2df135a5d9f6d2627bd125f362298430fdc06)
> but its way of affecting chain building changed later, maybe due my improvements done last year regarding the fix of #1418.

The ffd2df1 is master only. But with #1418 fix it was backported to 1.1.1 as well. Which effectively made this a regression on 1.1.1.

**Comment by DDvO** [comment]
Yes - the backporting commit was e2590c3a162eb118c36b09c2168164283aa099b4.

**Comment by DDvO** [comment]
> Is the mapping `rsassaPss undef rsaEncryption` in `obj_xref.txt` correct?

So far, this question of mine has been unanswered, but I am meanwhile pretty sure that this needs to be corrected to `rsassaPss undef rsassaPss`.

**Comment by t8m** [comment]
Please note that the RSA-PSS handling in current master is severally broken - see #13969.

I'd suggest for now to focus on fixing the regression issue on 1.1.1 and forward port it to master. Do not try to use master for development of the tests as it will produce bogus signatures.

**Comment by DDvO** [comment]
All right, so I've provided the original simper fix already discussed in #13968 as PR #13982.

**Comment by t8m** [comment]
The fix was merged to 1.1.1 branch.
